text/0000-build-observability.md

```diff
@@ -0,0 +1,358 @@
+              "parentSpanId": "",
+              "name": "buildpack-detect",
+              "startTimeUnixNano": "1581452772000000321",
+              "endTimeUnixNano": "1581452773000000789",
+              "droppedAttributesCount": 2,
+              "events": [
+                {
+                  "timeUnixNano": "1581452773000000123",
+                  "name": "detect-pass"
+                }
+              ],
+              "attributes": [
+                {
+                  "key": "buildpack-id",
+                  "value": {
+                    "stringValue": "heroku/nodejs-engine"
+                  }
+                }
+              ],
+              "droppedEventsCount": 1
+            }
+          ]
+        }
+      ]
+    }
+  ]
+}
+```
+
+
+### Buildpack telemetry files
+
+During a buildpack's `detect` or `build` execution, a buildpack may emit
+an OpenTelemetry File Export with tracing data to `/cnb/telemetry/#{buildpack-id}.jsonl`
+with contents like this:
+
+```json
+{
+  "resourceSpans": [
+    {
+      "resource": {
+        "attributes": [
+          {
+            "key": "lifecycle.version",
+            "value": {
+              "stringValue": "0.17.1"
+            }
+          }
+        ]
+      },
+      "scopeSpans": [
+        {
+          "scope": {},
+          "spans": [
+            {
+              "traceId": "",
+              "spanId": "",
+              "parentSpanId": "",
+              "name": "buildpack-detect",
+              "startTimeUnixNano": "1581452772000000321",
+              "endTimeUnixNano": "1581452773000000789",
+              "droppedAttributesCount": 2,
+              "events": [
+                {
+                  "timeUnixNano": "1581452773000000123",
+                  "name": "detect-pass"
+                }
+              ],
+              "attributes": [
+                {
+                  "key": "buildpack-id",
+                  "value": {
+                    "stringValue": "heroku/nodejs-engine"
+                  }
+                }
+              ],
+              "droppedEventsCount": 1
+            }
+          ]
+        }
+      ]
+    }
+  ]
+}
+```
+
+### Lifetime
+
+Telemetry files may be written at any point during the build, so that they
+are persisted in cases of failures to detect, failures to build, process
+terminations, or crashes. The `jsonl` format allows telemetry libraries to
+safely append additional json objects to the end of a telemetry file, so
+telemetry data can be flushed to the file frequently. Telemetry files should
+not be truncated or deleted so that telemetry processing by a platform can
+happen during or after a build. Telemetry files should not be included in the
+build result, as they are not relevant, and would likely negatively impact
+image size and reproduceability.
+
+### Access
+
+The telemetry files should be readable so that they may be analyzed by
+the user and/or platform. However, they should be write protected
```

Comment:
How will we accomplish this if buildpacks all run as the same user?

Reply:
During build, each buildpack runs synchronously, so we could prep the files/permissions prior to bin/build, so that each buildpack can only write to the expected files. 

During detect, many buildpacks will be running at the same time, and all under the same user. I can't think of a way to do block one buildpack from writing to another buildpack's detect telemetry.

To be fair, buildpacks can write to other buildpacks' `/layers` today. So maybe this section is a little overzealous. 

Maybe instead of blocking writes, we say something about the intent - buildpacks shall write only to their own telemetry files.